### PR TITLE
#3 Create Config form for custom module.

### LINKED ISSUE #3: #03 ­- Building Configuration forms
## Objective
At the end of this course, you will be able to

Create a very basic configuration form on a custom page
UProvide Default values for the configuration on the form
Save the configuration values on form submission.
## Exercise
Create a custom module that provides a configuration form available at the url “mymodule/config” to all users with the permission “administer content”
The form shall have 3 fields. Labels and Types of fields are arbitrary. Ensure you have diverse fields ­ Text, Select, Radio
Get the form to work such that the form values are saved and persisted on the form on reload.
The values submitted on the form should be accessible elsewhere on the same or a different module using the Config API.
## Bonus Exercise
Create your own permission and use that to restrict this page

### PR DESCRIPTION
**Pull request by amansrivastava**
#3

### CODE REVIEW

File: docroot/modules/custom/mymodule/mymodule.routing.yml [hunk 1]
```diff
@@ -0,0 +1,7 @@
+mymodule.custom_settings:
+  path: '/mymodule/config'
+  defaults:
+    _title: 'mymodule custom settings'
+    _form: 'Drupal\mymodule\Form\CustomSettingsForm'
+  requirements:
+    _permission: 'administer content'
```
Comment (joshirohit100) [inline docroot/modules/custom/mymodule/mymodule.routing.yml]: we should use the permission we created as part of this module - `administer mymodule configuration`

File: docroot/modules/custom/mymodule/src/Form/CustomSettingsForm.php [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+      '#type' => 'checkbox',
+      '#title' => $this->t('This is a checkbox example.'),
+      '#default_value' => $config->get('custom_checkbox'),
+    ];
+    $form['actions']['#type'] = 'actions';
+    $form['actions']['submit'] = [
+      '#type' => 'submit',
+      '#value' => $this->t('Save Custom Configs'),
+      '#button_type' => 'primary',
+    ];
```
Comment (joshirohit100) [inline docroot/modules/custom/mymodule/src/Form/CustomSettingsForm.php:47]: we dont add submit button but 
```
return parent::buildForm($form, $form_state);
```
and thus button added by parent class itself.

File: docroot/modules/custom/mymodule/src/Form/CustomSettingsForm.php [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+    return $form;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function validateForm(array &$form, FormStateInterface $form_state) {
+
+    parent::validateForm($form, $form_state);
+  }
```
Comment (joshirohit100) [inline docroot/modules/custom/mymodule/src/Form/CustomSettingsForm.php]: As this is optional and no validation is here, no need to add it here.

File: config/default/mymodule.settings.yml [hunk 4]
```diff
@@ -0,0 +1,3 @@
+example: example
```
Comment (joshirohit100) [inline config/default/mymodule.settings.yml]: we should as the `*.schema.yml` for this config in the module.

File: docroot/modules/custom/mymodule/src/Form/CustomSettingsForm.php (hunk 2 above)
Comment (amansrivastava) [reply docroot/modules/custom/mymodule/src/Form/CustomSettingsForm.php:47]: Hi @joshirohit100,
I added it to see change in Button Text.
I had `parent::buildForm($form, $form_state);` in return value earlier but I wanted to change the text of submit button and calling parent was overwriting my changes.

File: docroot/modules/custom/mymodule/src/Form/CustomSettingsForm.php (hunk 3 above)
Comment (amansrivastava) [reply docroot/modules/custom/mymodule/src/Form/CustomSettingsForm.php]: It is here for further development, I had a validation line to test this method.
I removed that validation to add something more reasonable. I'll be adding that in future implementations.

**Pull request merged**